src/main/kotlin/com/pravo/pravo/domain/promise/service/PromiseSettlementFacade.kt

```diff
@@ -0,0 +1,219 @@
+            throw UnauthorizedException(ErrorCode.UNAUTHORIZED, "모임장만 정산을 진행할 수 있습니다.")
+        }
+
+        return participants
+    }
+
+    private fun processSettlementWithValidation(
+        attendees: List<PromiseRole>,
+        absentees: List<PromiseRole>,
+        earnedPoint: Long,
+        promiseId: Long,
+        deposit: Int,
+    ) {
+        val cancelResults =
+            attendees.map { attendee ->
+                try {
+//                paymentService.cancelPayment(attendee.member.id, promiseId)
+                    true
+                } catch (e: Exception) {
+                    log.error("Failed to cancel payment for member ${attendee.member.id}: ${e.message}")
+                    throw IllegalStateException("결제 취소 실패로 인해 정산을 진행할 수 없습니다: ${e.message}")
+                }
+            }
+
+        if (cancelResults.all { it }) {
+            attendees.forEach { attendee ->
+                try {
+                    settlePoint(earnedPoint, attendee.member, promiseId)
+                } catch (e: Exception) {
+                    log.error("Failed to settle points for member ${attendee.member.id}: ${e.message}")
+                    throw IllegalStateException("포인트 정산 실패: ${e.message}")
+                }
+            }
+            absentees.forEach { fineLogService.saveFineLog(deposit.toLong(), it.member.id, promiseId) }
+        }
+    }
+
+    fun calculateEarnedPoint(
+        attendees: List<PromiseRole>,
+        absentees: List<PromiseRole>,
+        deposit: Int,
+    ): Long {
+        val totalPoint = absentees.size * deposit.toLong()
+        return totalPoint / attendees.size
+    }
+
+    fun settlePoint(
+        earnedPoint: Long,
+        member: Member,
+        promiseId: Long,
+    ) {
+        pointLogService.savePointLog(PointLogStatus.PLUS, earnedPoint, member.id, promiseId)
+        member.updatePoint(earnedPoint)
+    }
+
+    @Transactional
+    fun cancelPromise(
+        memberId: Long,
+        promiseId: Long,
+    ) {
+        val promiseRole = promiseRoleService.getPromiseRole(memberId, promiseId)
+        if (promiseRole.role == RoleStatus.ORGANIZER) {
+            throw UnauthorizedException(ErrorCode.UNAUTHORIZED, "모임장은 약속을 취소할 수 없습니다.")
+        }
+
+        paymentService.cancelPayment(memberId, promiseId)
+        promiseRole.updateStatus(ParticipantStatus.CANCELED)
+    }
+
+    private fun updateAndSplitParticipants(
+        participants: List<PromiseRole>,
+        attendedMemberIds: List<Long>,
+    ): Pair<List<PromiseRole>, List<PromiseRole>> {
+        return participants
+            .map { participant ->
+                val newStatus =
+                    if (attendedMemberIds.contains(participant.member.id)) {
+                        ParticipantStatus.ATTENDED
+                    } else {
+                        ParticipantStatus.NOT_ATTENDED
+                    }
+                participant.updateStatus(newStatus)
+                participant
```

Comment:
```suggestion
                participant.copy(status = newStatus)
```
P3: map 과정에서 사이드 이펙트를 피하기 위해 객체의 직접적인 업데이트 대신 새로운 형질을 갖는 객체를 반환하는건 어떨까요?

Reply:
좋은 지적 감사합니다! 하지만 copy를 쓰면 return type이 PromiseRole이 아니게 됩니다. 
copyWithStatus를 직접 구현해야 하는데 시간이 부족하여 이번 코멘트는 넘어가겠습니다 ㅜ